File: analytics/src/test/java/io/appmetrica/analytics/impl/referrer/service/ReferrerFromLibraryRetrieverTest.java
package io.appmetrica.analytics.impl.referrer.service;

import com.android.installreferrer.api.InstallReferrerClient;
import com.android.installreferrer.api.InstallReferrerStateListener;
import com.android.installreferrer.api.ReferrerDetails;
import io.appmetrica.analytics.coreapi.internal.executors.ICommonExecutor;
import io.appmetrica.analytics.impl.referrer.common.ReferrerInfo;
import io.appmetrica.analytics.testutils.CommonTest;
import org.junit.Before;
import org.junit.Test;
import org.junit.runner.RunWith;
import org.mockito.ArgumentCaptor;
import org.mockito.ArgumentMatcher;
import org.mockito.Captor;
import org.mockito.Mock;
import org.mockito.MockitoAnnotations;
import org.robolectric.RobolectricTestRunner;

import static org.mockito.ArgumentMatchers.argThat;
import static org.mockito.ArgumentMatchers.nullable;
import static org.mockito.Mockito.clearInvocations;
import static org.mockito.Mockito.mock;
import static org.mockito.Mockito.never;
import static org.mockito.Mockito.verify;
import static org.mockito.Mockito.verifyNoInteractions;
import static org.mockito.Mockito.when;

@RunWith(RobolectricTestRunner.class)
public class ReferrerFromLibraryRetrieverTest extends CommonTest {

    @Mock
    private InstallReferrerClient mClient;
    @Mock
    private ReferrerReceivedListener mListener;
    @Mock
    private ICommonExecutor executor;
    @Captor
    private ArgumentCaptor<Runnable> runnableCaptor;
    private ReferrerFromLibraryRetriever mReferrerFromLibraryRetriever;

    @Before
    public void setUp() throws Throwable {
        MockitoAnnotations.openMocks(this);
        mReferrerFromLibraryRetriever = new ReferrerFromLibraryRetriever(mClient, executor);
    }

    @Test
    public void testRetrieveReferrerOK() throws Throwable {
        final String installReferrer = "some test referrer";
        final long clickTimestamp = 438573450;
        final long installTimestamp = 3288994;
        ReferrerDetails details = mock(ReferrerDetails.class);
        when(details.getInstallReferrer()).thenReturn(installReferrer);
        when(details.getReferrerClickTimestampSeconds()).thenReturn(clickTimestamp);
        when(details.getInstallBeginTimestampSeconds()).thenReturn(installTimestamp);
        when(mClient.getInstallReferrer()).thenReturn(details);
        mReferrerFromLibraryRetriever.retrieveReferrer(mListener);
        ArgumentCaptor<InstallReferrerStateListener> referrerListenerCaptor = ArgumentCaptor.forClass(InstallReferrerStateListener.class);
        verify(mClient).startConnection(referrerListenerCaptor.capture());
        referrerListenerCaptor.getValue().onInstallReferrerSetupFinished(0);
        verify(mClient).endConnection();
        verifyNoInteractions(mListener);
        clearInvocations(mClient);
        verify(executor).execute(runnableCaptor.capture());
        runnableCaptor.getValue().run();
        verify(mClient, never()).getInstallReferrer();
        verify(mListener).onReferrerReceived(argThat(new ArgumentMatcher<ReferrerInfo>() {
            @Override
            public boolean matches(ReferrerInfo argument) {
                return argument.installReferrer.equals(installReferrer) &&
                        argument.installBeginTimestampSeconds == installTimestamp &&
                        argument.referrerClickTimestampSeconds == clickTimestamp &&
                        argument.source == ReferrerInfo.Source.GP;
            }
        }));
        verify(mListener, never()).onReferrerRetrieveError(nullable(Throwable.class));
    }

    @Test
    public void testRetrieveReferrerThrows() throws Throwable {
        RuntimeException exception = new RuntimeException();
        when(mClient.getInstallReferrer()).thenThrow(exception);
        mReferrerFromLibraryRetriever.retrieveReferrer(mListener);
        ArgumentCaptor<InstallReferrerStateListener> referrerListenerCaptor = ArgumentCaptor.forClass(InstallReferrerStateListener.class);
        verify(mClient).startConnection(referrerListenerCaptor.capture());
        referrerListenerCaptor.getValue().onInstallReferrerSetupFinished(0);
        verify(mClient).endConnection();
        verifyNoInteractions(mListener);
        verify(executor).execute(runnableCaptor.capture());
        runnableCaptor.getValue().run();
        verify(mListener, never()).onReferrerReceived(nullable(ReferrerInfo.class));
        verify(mListener).onReferrerRetrieveError(exception);
    }

    @Test
    public void testRetrieverReferrerBadCode() throws Throwable {
        when(mClient.getInstallReferrer()).thenReturn(mock(ReferrerDetails.class));
        mReferrerFromLibraryRetriever.retrieveReferrer(mListener);
        ArgumentCaptor<InstallReferrerStateListener> referrerListenerCaptor = ArgumentCaptor.forClass(InstallReferrerStateListener.class);
        verify(mClient).startConnection(referrerListenerCaptor.capture());
        referrerListenerCaptor.getValue().onInstallReferrerSetupFinished(1);
        verify(mClient).endConnection();
        verifyNoInteractions(mListener);
        verify(executor).execute(runnableCaptor.capture());
        runnableCaptor.getValue().run();
        verify(mListener, never()).onReferrerReceived(nullable(ReferrerInfo.class));
        verify(mListener).onReferrerRetrieveError(argThat(new ArgumentMatcher<Throwable>() {
            @Override
            public boolean matches(Throwable argument) {
                return argument.getMessage().equals("Referrer check failed with error 1");
            }
        }));
    }
}

File: analytics/src/test/java/io/appmetrica/analytics/impl/referrer/service/ReferrerFromLibraryRetrieverTest.kt
package io.appmetrica.analytics.impl.referrer.service

import android.content.Context
import com.android.installreferrer.api.InstallReferrerClient
import com.android.installreferrer.api.InstallReferrerStateListener
import com.android.installreferrer.api.ReferrerDetails
import io.appmetrica.analytics.coreapi.internal.executors.ICommonExecutor
import io.appmetrica.analytics.testutils.CommonTest
import io.appmetrica.analytics.testutils.on
import io.appmetrica.analytics.testutils.staticRule
import org.junit.Rule
import org.junit.Test
import org.junit.runner.RunWith
import org.mockito.kotlin.argThat
import org.mockito.kotlin.argumentCaptor
import org.mockito.kotlin.doReturn
import org.mockito.kotlin.inOrder
import org.mockito.kotlin.mock
import org.mockito.kotlin.whenever
import org.robolectric.RobolectricTestRunner

@RunWith(RobolectricTestRunner::class)
class ReferrerFromLibraryRetrieverTest : CommonTest() {

    private val context: Context = mock()
    private val listener: ReferrerReceivedListener = mock()
    private val executor: ICommonExecutor = mock()
    private val runnableCaptor = argumentCaptor<Runnable>()

    private val installReferrer = "some test referrer"
    private val clickTimestamp: Long = 438573450
    private val installTimestamp: Long = 3288994

    private val installReferrerDetails = mock<ReferrerDetails> {
        on { installReferrer } doReturn installReferrer
        on { referrerClickTimestampSeconds } doReturn clickTimestamp
        on { installBeginTimestampSeconds } doReturn installTimestamp
    }

    private val client: InstallReferrerClient = mock {
        on { installReferrer } doReturn installReferrerDetails
    }

    private val referrerListenerCaptor = argumentCaptor<InstallReferrerStateListener>()

    private val installReferrerClientBuilder: InstallReferrerClient.Builder = mock {
        on { build() } doReturn client
    }

    @get:Rule
    val installReferrerClientRule = staticRule<InstallReferrerClient> {
        on { InstallReferrerClient.newBuilder(context) } doReturn installReferrerClientBuilder
    }

    private val referrerFromLibraryRetriever by setUp { ReferrerFromLibraryRetriever(context, executor) }

    @Test
    fun `retrieveReferrer if ok`() {
        referrerFromLibraryRetriever.retrieveReferrer(listener)
        inOrder(client, executor, listener) {
            verify(client).startConnection(referrerListenerCaptor.capture())
            verifyNoMoreInteractions()
            referrerListenerCaptor.firstValue.onInstallReferrerSetupFinished(0)
            verify(executor).execute(runnableCaptor.capture())
            verifyNoMoreInteractions()
            runnableCaptor.firstValue.run()
            verify(listener).onReferrerReceived(argThat {
                this.installReferrer == installReferrer &&
                    this.referrerClickTimestampSeconds == clickTimestamp &&
                    this.installBeginTimestampSeconds == installTimestamp
            })
            verify(client).endConnection()
            verifyNoMoreInteractions()
        }
    }

    @Test
    fun `retrieveReferrer if thrown`() {
        val exception = RuntimeException()
        whenever(client.installReferrer).thenThrow(exception)
        referrerFromLibraryRetriever.retrieveReferrer(listener)
        inOrder(client, executor, listener) {
            verify(client).startConnection(referrerListenerCaptor.capture())
            verifyNoMoreInteractions()
            referrerListenerCaptor.firstValue.onInstallReferrerSetupFinished(0)
            verify(executor).execute(runnableCaptor.capture())
            verifyNoMoreInteractions()
            runnableCaptor.firstValue.run()
            verify(listener).onReferrerRetrieveError(exception)
            verify(client).endConnection()
            verifyNoMoreInteractions()
        }
    }

    @Test
    fun `retrieveReferrer if bad code`() {
        referrerFromLibraryRetriever.retrieveReferrer(listener)
        inOrder(client, executor, listener) {
            verify(client).startConnection(referrerListenerCaptor.capture())
            verifyNoMoreInteractions()
            referrerListenerCaptor.firstValue.onInstallReferrerSetupFinished(1)
            verify(executor).execute(runnableCaptor.capture())
            verifyNoMoreInteractions()
            runnableCaptor.firstValue.run()
            verify(listener).onReferrerRetrieveError(argThat {
                this is IllegalStateException && this.message == "Referrer check failed with error 1"
            })
            verifyNoMoreInteractions()
        }
    }
}
